### PR TITLE
Protocol improvements (Pool Proxy, finalizeTransfer, collateralizedBalance, AssetType, DynamicConfig)

### PR DESCRIPTION
**Pull request by WalidOfNow**
## Security Checklist

- [ ] 1. Re-Entrancy
- [ ] 2. Arithmetic Over/Under Flows
- [ ] 3. Unexpected Ether
- [ ] 4. Delegatecall
- [ ] 5. Default Visibilities
- [ ] 6. Entropy Illusion
- [ ] 7. External Contract Referencing
- [ ] 8. Short Address/Parameter Attack (off chain)
- [ ] 9. Unchecked CALL Return Values
- [ ] 10. Race Conditions / Front Running
- [ ] 11. Denial Of Service (DOS)
- [ ] 12. Block Timestamp Manipulation
- [ ] 13. Constructors with Care
- [ ] 14. Uninitialized Storage Pointers
- [ ] 15. Floating Points and Precision
- [ ] 16. Tx.Origin Authentication
- [ ] 17. Address.isContract Re-Entrancy via Constructor

### ⚠️ NOTES ⚠️

Make sure to think about each of these exploits in this PR.

### CODE REVIEW

File: contracts/protocol/tokenization/base/MintableIncentivizedERC721.sol [hunk 1]
```diff
@@ -642,15 +642,16 @@ abstract contract MintableIncentivizedERC721 is
     function _transferCollaterizable(
         address from,
         address to,
-        uint256 tokenId,
-        bool isUsedAsCollateral
-    ) internal virtual {
-        MintableIncentivizedERC721._transfer(from, to, tokenId);
+        uint256 tokenId
+    ) internal virtual returns (bool isUsedAsCollateral) {
+        isUsedAsCollateral = _isUsedAsCollateral[tokenId];
```
Comment (GopherJ) [inline contracts/protocol/tokenization/base/MintableIncentivizedERC721.sol:647]: what if from == to?

File: contracts/protocol/pool/PoolCore.sol [hunk 2]
```diff
@@ -656,6 +624,37 @@ contract PoolCore is
...
+        SupplyLogic.executeFinalizeTransferERC721(
+            _reserves,
+            _reservesList,
+            _usersConfig,
+            DataTypes.FinalizeTransferParams({
+                asset: asset,
+                from: from,
+                to: to,
+                usedAsCollateral: usedAsCollateral,
+                value: 1,
```
Comment (GopherJ) [inline contracts/protocol/pool/PoolCore.sol:649]: value is too general I think we can rename to amount

File: contracts/protocol/tokenization/base/MintableIncentivizedERC721.sol (hunk 1 above)
Comment (GopherJ) [reply contracts/protocol/tokenization/base/MintableIncentivizedERC721.sol:647]: also need to change _liquidateNTokens if we really want to do so

File: contracts/protocol/libraries/logic/SupplyLogic.sol [hunk 3]
```diff
@@ -446,6 +435,56 @@ library SupplyLogic {
...
+                        reservesList,
+                        usersConfig[params.from],
+                        params.asset,
+                        params.from,
+                        params.reservesCount,
+                        params.oracle
+                    );
+                }
+
+                if (params.balanceFromBefore == 1) {
```
Comment (GopherJ) [inline contracts/protocol/libraries/logic/SupplyLogic.sol:477]: executeCollateralizedERC721 also needs to be updated

Comment (GopherJ) [reply contracts/protocol/libraries/logic/SupplyLogic.sol:477]: ok maybe not necessary, finalizeTransfer only useful for transfer

File: contracts/protocol/tokenization/base/MintableIncentivizedERC721.sol [hunk 4]
```diff
@@ -642,15 +642,16 @@ abstract contract MintableIncentivizedERC721 is
...
 
-        if (isUsedAsCollateral) {
+        if (from != to && isUsedAsCollateral) {
             _userState[from].collaterizedBalance -= 1;
```
Comment (zhoujia6139) [inline contracts/protocol/tokenization/base/MintableIncentivizedERC721.sol:650]: Maybe we can revert directly if `from == to`, but it's ok, it's on your decide.

**Review by zhoujia6139 — APPROVED** [review]

**Pull request merged**